brainglobe_stitch/tile.py

```diff
@@ -55,3 +55,148 @@ def __init__(
         self.tile_id: int = int(attributes["tile"])
         self.illumination_id: int = int(attributes["illumination"])
         self.angle: float = float(attributes["angle"])
+
+
+class Overlap:
+    """
+    Overlap class to store information about the overlap between two tiles.
+
+    Attributes
+    ----------
+    coordinates : npt.NDArray
+        The coordinates for the start of the overlap in the fused image.
+    size : List[npt.NDArray]
+        The size of the overlap in the fused image for each resolution pyramid.
+    tiles : Tuple[Tile, Tile]
+        The overlapping tiles.
+    local_coordinates : List[Tuple[npt.NDArray, npt.NDArray]]
+        A list of tuples representing the starting coordinates for the overlap
+        in each tile's data pyramid for each resolution pyramid level.
+        The tuple contains the starting coordinates for the overlap in each
+        tile.
+    """
+
+    def __init__(
+        self,
+        overlap_coordinates: npt.NDArray,
+        overlap_size: npt.NDArray,
+        tile_i: Tile,
+        tile_j: Tile,
+    ):
+        """
+        Initialize the overlap.
+
+        Parameters
+        ----------
+        overlap_coordinates: npt.NDArray
+            The starting coordinates for the overlap in the fused image.
+        overlap_size: npt.NDArray
+            The size of the overlap in the fused image [z,y,x].
+        tile_i: Tile
+            The first tile.
+        tile_j: Tile
+            The second tile.
+        """
+        self.coordinates: npt.NDArray = overlap_coordinates
+        self.size: List[npt.NDArray] = [overlap_size]
+        self.tiles: Tuple[Tile, Tile] = (tile_i, tile_j)
+        self.local_coordinates: List[Tuple[npt.NDArray, npt.NDArray]] = [
+            (
+                np.zeros(self.coordinates.shape),
+                np.zeros(self.coordinates.shape),
+            ),
+        ]
+        self.get_local_overlap_coordinates()
+
+    def get_local_overlap_coordinates(self) -> None:
+        """
+        Calculate the starting coordinates for the overlap in
+        each tile's data pyramid.
+        """
+        tile_shape = self.tiles[0].data_pyramid[0].shape
+        resolution_pyramid = self.tiles[0].resolution_pyramid
+
+        for i in range(self.coordinates.shape[0]):
+            if self.tiles[0].position[i] < self.tiles[1].position[i]:
```

Comment:
Why are we referring to the first and second tile? 🤔

Reply:
I am understanding that the list of tiles is calculated once outside the creation of the `Overlap` object, and so they will have fixed positions... how do the first and second tile relate to the local coords of any overlap?

Reply:
Each `Overlap` is made up of two tiles. There are two separate but connected sets of coordinates that we keep track of, the "global" coordinates representing where the overlap lies in the final fused image, and the "local" coordinates which hold information on where the overlap lies in the `Tile` it originates in. The `get_local_overlap_coordinates` method checks which tile is "first" in each dimension to see if the overlap is at the beginning or the end for that specific dimension. This later allows us to extract, and modify the data stored in each overlap prior to fusing the entire image.

Reply:
Oh wow I would have never guessed that 😂